kong/pdk/response.lua

```diff
@@ -517,14 +519,94 @@ local function new(self, major_version)
   end
 
 
-  --function _RESPONSE.set_raw_body(body)
-  --  -- TODO: implement, but how?
-  --end
+  ---
+  -- Returns the full body when the last chunk has been read
+  --
+  -- @function kong.response.get_raw_body
+  -- @phases `body_filter`
+  -- @treturn string body The full body when the last chunk has been read,
+  --                      otherwise returns `nil`
+  -- @usage
+  -- local body = kong.response.get_raw_body()
+  -- if body then
+  --   ...
+  -- end
+  function _RESPONSE.get_raw_body()
+    check_phase(PHASES.body_filter)
+
+    local body_buffer
+    local chunk = arg[1]
+    local eof = arg[2]
+    if eof then
+      body_buffer = self.ctx.core.body_buffer
+      if not body_buffer then
+        return chunk
+      end
+    end
+
+    if type(chunk) == "string" and chunk ~= "" then
+      if not eof then
+        body_buffer = self.ctx.core.body_buffer
+      end
+
+      if body_buffer then
+        local n = body_buffer.n + 1
+        body_buffer.n = n
+        body_buffer[n] = chunk
+
+      else
+        body_buffer = {
+          chunk,
+          n = 1
+        }
+
+        self.ctx.core.body_buffer = body_buffer
+      end
+    end
+
+    if eof then
+      if body_buffer then
+        body_buffer = concat(body_buffer, "", 1, body_buffer.n)
+      else
+        body_buffer = ""
+      end
+
+      arg[1] = body_buffer
+      return body_buffer
+    end
+
+    arg[1] = nil
+    return nil
+  end
+
+
+  ---
+  -- Sets the body of the response
   --
+  -- The `body` argument must be a string and will not be processed in any way.
+  -- This function cannot anymore change the `Content-Length` header if one was
+  -- added. So if you decide to use this function, the `Content-Length` header
+  -- should also be cleared, e.g. in `header_filter` phase.
   --
-  --function _RESPONSE.set_body(args, mimetype)
-  --  -- TODO: implement, but how?
-  --end
+  -- @function kong.response.set_raw_body
+  -- @phases `body_filter`
+  -- @tparam string body The raw body
+  -- @return Nothing; throws an error on invalid inputs.
+  -- @usage
+  -- kong.response.set_raw_body("Hello, world!")
+  function _RESPONSE.set_raw_body(body)
```

Comment:
Although the implementationf of this function does allow it to be called multiple times, but
if I understand correctly, it would be best practice for a user to use combination of get_raw_body and then set_raw_body. And the function name indicates it's setting the body
not chunk. Does it make sense to limit user to only be able to call this function once? By having this limitation, user may have less confusing on how body_filter works and when/how
 to use those two functions.

Reply:
@fffonion,

Let's say there are two plugins:
- plugin 1. uppercase response, priority 1
- plugin 2. lowercase response, priority 2

and both plugins are added to route and executed.

1. both call get_raw_body
2. plugin 1 gets full body and turns the response to uppercase and calls set_raw_body
3. plugin 2 gets full body set by plugin 1 and turns the response to lowercase and calls set_raw_body

I think in this scenario it is beneficial to be able to call it multiple times. Of course the above example is crafted, but perhaps there are real use cases, and I think it is strange if plugin 1 can set the body but the plugin 2 cannot anymore.

Reply:
I see, in that case this makes more sense to me.

Reply:
@fffonion they don't need to be called as pairs. On last chunk we don't clear anything. On chunks except the last, calling `get_raw_body` will set `ngx.var[1]` to `nil` (this does not happen on last chunk). I think they can be called independently (perhaps there are still some caveats).

This would be wrong usage in plugin:
```
if ngx.var[2] then
  local body = kong.response.get_raw_body()
end
```

As it is too late to start buffering on last chunk, and thus it possibly only returns the last chunk (returns full body if some other plugin has called get_raw_body unconditionally). I am not sure how to tackle this, or does it need to be? Can we just call it invalid use?

Reply:
On my testing `body_filter` is usually called at least `2` times. When `ngx.arg[2] == true` the `ngx.arg[1] == nil` seems to be what happens there.

Reply:
@bungle It's okay to have various use case and I don't think the approach here is techically wrong. I'm just more
concerned about the comment (which will become docs automatically), and we should give user a clear idea on
what's happening (like what this function will affect) to avoid misuse, especially those doesn't have prior knowledge to openresty or have written the normal body_filter code but haven't read ours.
I'm also okay to leave this as a future improvement since it's not a blocker after all.